### ISSUE TITLE
Audacity crashes at startup (MinGW and OpenBSD)

### ISSUE THREAD

**Issue by carlo-bramini** [audacity/audacity#495]
I successfully compiled Audacity for Windows by using CMake and MinGW.
Unfortunately, I'm getting a crash after the application started.
After running, the program shows the small splash dialog for about one
second and then it appears a messagebox with this error:

```
Audacity:
*** Caught unhandled unknown exception, terminating.
```

After some debugging, I discovered that the program crashes at this
line, inside src/ProjectManager.cpp:

https://github.com/audacity/audacity/blob/866edf88011b97bf4c2d5a6dda49374ccaa602e5/src/ProjectManager.cpp#L536

This error has been discussed a bit in the mailing list and it seems to happen also on OpenBSD.
On Windows, it seems that only GCC is affected, because I also configured for compiling with VisualStudio 2019 and the resulting executable was not crashing.

**Comment by carlo-bramini** [comment]
For completeness, I'm also adding the backtrace generated from OpenBSD and posted by an user to  the mailing list.
[openbsd-backtrace.txt](https://github.com/audacity/audacity/files/4559981/openbsd-backtrace.txt)

**Comment by carlo-bramini** [comment]
Some recent fixes made possible to build for Cygwin and the same issue happens on that environment.

**Comment by lllucius** [comment]
Are you able to build a debug version?  Can you get a backtrace?

**Comment by carlo-bramini** [comment]
Please excuse my ignorance, how do I build and I do this backtrace?

**Comment by lllucius** [comment]
Try adding a build type to the cmake command:

cmake -DCMAKE_BUILD_TYPE=Debug <any other parameters>

What that's going to do is turn of optimization.  If my guess is right, that will get you past the crash.

Of course you don't want to run that way, so if it does get you past the crash, remove the build type and try inserting the following line at the very top of src/ProjectWindow.h, src/Project.h and src/menus/ViewMenus.cpp.

#include <wx/wx.h>

Let me know how you get along.

**Comment by lllucius** [comment]
BTW:  Do you have a full set of patches I could try here.  I'd also like to try recreating this crash.  I think it's very similar to another we're having and this might shed a bit more light on the cause.

**Comment by lllucius** [comment]
Carlo, if your problem is the same as another that we've been working on, the I believe we have an answer for you.  This may or may not fix your problem, but it certainly fixed the other problem and I'm thinking that it should fix yours as well.

We tracked down the root cause of the problem to this wxWidgets bug:

https://trac.wxwidgets.org/ticket/15884
wxWidgets/wxWidgets@72aad0d

I backported their fix to wxWidgets 3.0.5.1 so I could verify that it does indeed correct the issue on Linux.

weakref.patch.gz

I've also included the patched weakref.h so it should be easy for you to test as well.

**Comment by carlo-bramini** [comment]
> Try adding a build type to the cmake command:
> 
> cmake -DCMAKE_BUILD_TYPE=Debug
> 
I configured and I compiled the sources in the way you suggested me, but unfortunately nothing changed.

**Comment by lllucius** [comment]
Thanks for trying, but did you happen to see my last post about the wxWidgets bug?  i think that may be the problem.

**Comment by carlo-bramini** [comment]
> BTW: Do you have a full set of patches I could try here. I'd also like to try recreating this crash. I think it's very similar to another we're having and this might shed a bit more light on the cause.

I'm compiling Audacity in 4 ways:
1) CYGWIN build, it uses wxGTK and POSIX layer.
2) Native Windows build, it uses CYGWIN for shell and MinGW-w64 cross-compiler.
3) Native Windows build, it uses MSYS2 for shell and MinGW-w64 cross-compiler.
4) Native Windows build, it uses Visual Studio 2019 Community Edition.

The builds (1) and (2) use cmake with Unix Makefile generator.
The (1) requires PR #447, #467, #556 and #557 for compiling correctly everything.
Unfortunately, something strange happens when doing the final link of Audacity, because you will see that an unknown error is printed on the console. In my opinion, this happens because there is a bug in CMake. As you can see here:
https://gitlab.kitware.com/cmake/cmake/-/blob/master/Modules/Platform/CYGWIN-GNU.cmake#L44
CMake adds `"-Wl,--export-all-symbols"` option when linking, which puts all symbols into the import library, but this is bad, at least for Audacity.
The Portable Executable format allows to export a maximum number of 65535 symbols.
Audacity overflows this limit and GNU linker returns an error because of this.
Using all system libraries instead of local ones was not enough for recovering this error, so the only solution is to enter into the build directory, open `src/CMakeFiles/Audacity.dir/link.txt` and remove the above option. After that, you can launch `make` again and the compilation will complete successfully.

The builds (2) and (3) requires few, small fixes to lv2, portmixer and portsmf. The native builds for Windows also raises some strange messages when building the resource file. I have identified the problem and I collected the fixes in this branch:
https://github.com/carlo-bramini/audacity/tree/rsrc_fixes_1
But I'm having some troubles to make a PR of this because GITHUB gave to me this message:
> We’ve detected the file encoding as ISO-8859-1. When you commit changes we will transcode it to UTF-8.

As you can see, in my branch some strings of win/audacity.rc were corrupted after the commit and I do not understand very well the reasons. The build (3) uses MSYS2 as shell and "MSYS Makefiles" as generator and, when linking, the problem is the opposite of the one I'm getting with "Unix Makefiles" under Cygwin. In other words, `set( CMAKE_ENABLE_EXPORTS ON )` does not seem to be respected by CMake because import library `libaudacity.dll.a` is not created. As result, the executable is created, but the modules fail to build since the library is missing. Here the only solution that I have found is to create `audacity.def` file with pexports or other similar tool, create `libaudacity.dll.a` with `dlltool` and then launch again `make` to complete the build.

I hope that you will find this useful.

**Comment by Paul-Licameli** [comment]
It is not a wxWidgets issue.

I fixed a similar thing at commit a085057b6616d99c6b19e48b5905e737dd636d7d

But from your stack trace, I see other work is needed.

PRL

**Comment by Paul-Licameli** [comment]
To explain more, and why I don't think it's related to wxWidgets issues.

The ClientData thing stores a table of factory functions and invokes all of
them in BuildAll().  There is more than one instantiation of the class
template ClientData.  The reason this exists is to break compilation and
linkage dependencies, so that the class AudacityProject can be low in the
dependency graph but the various cooperating parts of the project can be
higher-level.

The several cooperating parts insert factory functions in the static table
of factories traversed by BuildAll(), and that insertion happens at the
time of initialization of file-scope static objects.

But as you may know, in C++ the relative ordering of initialization of
file-scope static objects in different compilation units is not specified.
I think it should be predictable with a given platform and build system,
but may vary between build systems.

So the sequence of the factories in ClientData may vary, and there may have
been a hidden ordering dependency among them, which happened not be
violated on any other platform yet that we regularly build, but you managed
to build in a way violating that ordering.

That's the sort of problem I discovered and fixed at
a085057b6616d99c6b19e48b5905e737dd636d7d, but there may be other
dependencies.

I mean to find and fix all such dependencies now.  There are 30 factories
in the relevant table, and I only need to test 30 cyclic permutations of
the table (not all 30-factorial!) to make sure I test all pairs of
functions each way.

PRL

**Comment by Paul-Licameli** [comment]
Carlo, try the simple fix at commit 4ac96376bb6f5832e735e08314e930e9d8ae926d

Paul

**Comment by Paul-Licameli** [comment]
I have now completed that testing I described, of 30 cyclic permutations of
the sequence of invocations of the factories, on one of the platforms
(Mac), so I'm satisfied there are not other such surprises waiting with
other variations of the build system.

PRL

**Comment by rrrapha** [comment]
https://github.com/audacity/audacity/commit/4ac96376bb6f5832e735e08314e930e9d8ae926d fixes the issue for me on OpenBSD ;)

**Comment by carlo-bramini** [comment]
> 
> Carlo, try the simple fix at commit [4ac9637](https://github.com/audacity/audacity/commit/4ac96376bb6f5832e735e08314e930e9d8ae926d)

IT WORKS!
Thank you very much!

BTW, me too, I have a little news for you: Audacity, if compiled with MinGW, works again on Windows XP and perhaps also on NT. And this is the target that I wanted to get ;)

![Windows-XP](https://user-images.githubusercontent.com/30959007/84536946-28ce6e80-acef-11ea-8097-51e3300a4777.png)

**Comment by JamesCrook** [comment]
Resolve fixed
> IT WORKS!
